core/drivers/stm32mp15_huk.c

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2022, Foundries.io Limited
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <crypto/crypto.h>
+#include <drivers/stm32_bsec.h>
+#include <kernel/tee_common_otp.h>
+#include <mempool.h>
+#include <platform_config.h>
+#include <string.h>
+#include <string_ext.h>
+
+static bool stm32mp15_huk_initialized;
+
+static TEE_Result stm32mp15_read_uid(uint32_t *uid)
+{
+	uint32_t *q = uid;
+
+	if (stm32_bsec_shadow_read_otp(q++, UID0_OTP))
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_shadow_read_otp(q++, UID1_OTP))
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_shadow_read_otp(q++, UID2_OTP))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32mp15_read_otp(uint32_t otp, uint32_t *key, bool *locked)
+{
+	bool tmp = false;
+
+	if (stm32_bsec_read_permanent_lock(otp, &tmp))
+		return TEE_ERROR_GENERIC;
+
+	if (tmp && stm32_bsec_read_sw_lock(otp, &tmp))
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_shadow_read_otp(key, otp))
+		return TEE_ERROR_GENERIC;
+
+	*locked = *locked && tmp;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ *  AES-GCM: nonce must be unique per message and key.
+ *
+ *  This function always uses the same key - once its locked - with the same
+ *  unique message hence the nonce can be any constant.
+ */
+static TEE_Result aes_gcm_encrypt_uid(uint8_t *key, size_t key_len,
+				      uint8_t *out, size_t *out_len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	const uint8_t nonce[12] = { 0x55 };
+	uint32_t uid[4] = { 0 };
+	uint8_t tag[16] = { 0 };
+	size_t nonce_len = sizeof(nonce);
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	void *ctx = NULL;
+
+	ret = stm32mp15_read_uid(uid);
+	if (ret)
+		goto err;
+
+	ret = crypto_authenc_alloc_ctx(&ctx, TEE_ALG_AES_GCM);
+	if (ret)
+		goto err;
+
+	ret = crypto_authenc_init(ctx, TEE_MODE_ENCRYPT, key, key_len, nonce,
+				  nonce_len, TEE_AES_BLOCK_SIZE, 0, uid_len);
+	if (ret)
+		goto auth_err;
+
+	ret = crypto_authenc_enc_final(ctx, (uint8_t *)uid, sizeof(uid),
+				       out, out_len, tag, &tag_len);
+	if (ret)
+		goto auth_err;
+
+	crypto_authenc_final(ctx);
+auth_err:
+	crypto_authenc_free_ctx(ctx);
+err:
+	memzero_explicit(uid, uid_len);
+	memzero_explicit(tag, tag_len);
```

Comment:
`uid` is not sensitive ,no need for memzeor_explicit.
I don't think `tag` is neither.

Reply:
From the SoC spec, UID fuses are in the "non-secure" fuses area, meaning non-secure world can read it any time.
For the tag, I don't think it can evade information from derived key.

As for the zynqmp drier, sorry i did not catch this issue. 
Not a big deal here, but let's have code that do what's needded and don't do what is not.